### ISSUE TITLE
Eager execution support ?

### ISSUE THREAD

**Issue by nakamichiworks** [rstudio/tensorflow#196]
Tensorflow is planned to support eager execution mode: [Eager Execution: An imperative, define-by-run interface to TensorFlow](https://research.googleblog.com/2017/10/eager-execution-imperative-define-by.html).

This might be useful for R users of Tensorflow, but it is intended to be used with pythonic constructs such as annotations (for costum gradients) and class inheritance (to define models).

Do you have any plan or idea to make eager execution easy to use for R users ?

**Comment by jjallaire** [comment]
I expect that a lot of eager execution would already work well with the R interface. The annotated functions however would require extra work. In terms of class inheritance to define models, I don't know if that's strictly required. If it is then we could either find another way to accomplish the same thing use R6 objects (we already do that in Keras for defining custom layers and callbacks).

**Comment by nakamichiworks** [comment]
I have tried some of eager-mode tutorials in R ([tfe-guide-R.md](https://gist.github.com/nakamichiworks/9cd6f9a127a8637f2e55a536028685cd)), and almost all functions have worked with no problems. 

Some thoughts on the R interface:

- In R, you can just call Python's annotation function and pass a function definition to it. It seems syntactically natural to me and we might not need extra syntax for the annotations.

```r
# annotation example
log1pexp = tfe$custom_gradient(
  function(x) {
    e = tf$exp(x)
    grad = function(dy) {
      dy * (1 - 1 / (1 + e))
    }
    list(tf$log(1 + e), grad)
  })
```

- Extending [`tfe.Network`](https://www.tensorflow.org/versions/master/api_docs/python/tf/contrib/eager/Network) to define models is not strictly necessary but useful abstraction. However, using R6 to mimic the behavior of `tfe.Network` might be cumbersome because `tfe.Network` inherits `tf.layers.Layer` and is expected to work well with existing TF functions. I don't know it is worth developing many R wrapper functions to use the class.

I hope we could find more suitable way for R to use imperative interfaces to deep learning frameworks, but have no good idea for now.

**Comment by jjallaire** [comment]
Great to hear that things mostly work as expected!

Do the eager mode tensors get a special class? If so we could create S3 methods for `as.array`, `as.matrix` etc. to convert them into R native objects.

**Comment by nakamichiworks** [comment]
Yes, eager mode tensors have the special class for them, and I think it would be very good to have `as.array()` and `as.matrix()` methods instead of `EagerTensor$numpy()` !

- Eager mode:
```r
library(tensorflow)
tfe = tf$contrib$eager
tfe$enable_eager_execution()
x = tf$constant(0)
print(class(x))
#> [1] "python.builtin.EagerTensor"                      
#> [2] "tensorflow.python.framework.ops._EagerTensorBase"
#> [3] "tensorflow.python.framework.ops.Tensor"          
#> [4] "tensorflow.python.framework.ops._TensorLike"     
#> [5] "python.builtin.object"```
```

- Non-eager mode:
```r
library(tensorflow)
x = tf$constant(0)
print(class(x))
#> [1] "tensorflow.python.framework.ops.Tensor"     
#> [2] "tensorflow.python.framework.ops._TensorLike"
#> [3] "python.builtin.object"
```

On the R interface, I have just heared that [Keras will support dynamic graphs with TF eager backend](https://twitter.com/fchollet/status/931026435715641344). We might be able to expect to rely on the Keras interface.

**Comment by jjallaire** [comment]
Here's some initial work: https://github.com/rstudio/tensorflow/commit/dadbc96941bceda7573488e49fb75ad77f5ef6b0

**Comment by jjallaire** [comment]
FYI I renamed with a `tfe_` prefix here in anticipation of export other functions from `tf$contrib$eager`: https://github.com/rstudio/tensorflow/commit/a76cc5ce989d1a3956a68ac45989029778a05e7b
